### PR TITLE
Optionally provide ca_pin as a file path

### PR DESCRIPTION
**Pull request by GavinFrazar**
This adds support for specifying ca_pin as a file path (or list of file paths): #12444

I followed the example we use for reading auth_token from a file path, reused some functionality from there and did some refactoring to handle paths in a cross-platform way (in case we ever bring the teleport daemon to windows).

### CODE REVIEW

File: lib/service/cfg.go [hunk 1]
```diff
@@ -276,6 +276,28 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
 	return false, nil
 }
 
+// ApplyCAPins assigns the given CA pin(s), but only if each pin is not empty.
+func (cfg *Config) ApplyCAPins(caPins []string) error {
```
Comment (marcoandredinis) [inline lib/service/cfg.go:280]: I think we can remove the second loop and the `if caPins[i] != ""` because `filepath.IsAbs(value)` should handle that corner case
What do you think?

**Review by marcoandredinis — COMMENTED** [review]
Can we please add a test?

This seems to be a simple change, but I think we should add a test nevertheless

File: lib/service/cfg.go (hunk 1 above)
Comment (GavinFrazar) [reply lib/service/cfg.go:280]: TryReadFileAsValue will just return the string if it doesn't look like an absolute path. I think we should ignore blank lines in case a user has something like this in their config:

```
  ca_pin:
    - /Users/gavin/foo
    -
```

We were already doing the same thing for auth_token. I could condense it to one loop though, like this:

```go
	filteredPins := make([]string, 0, len(caPins))
	for _, pin := range caPins {
		if pin != "" {
			var err error
			pin, err = utils.TryReadValueAsFile(pin)
			if err != nil {
				return trace.Wrap(err)
			}
			filteredPins = append(filteredPins, pin)
		}
	}
```

Comment (marcoandredinis) [reply lib/service/cfg.go:280]: Looks good :+1:

File: lib/config/fileconf.go [hunk 2]
```diff
@@ -778,7 +778,8 @@ type StaticToken string
 // apart and constructs a services.ProvisionToken which contains the token,
 // role, and expiry (infinite).
 func (t StaticToken) Parse() (*types.ProvisionTokenV1, error) {
-	parts := strings.Split(string(t), ":")
+	// Split only on the first ':', for future cross platform compat with windows paths
```
Comment (codingllama) [inline lib/config/fileconf.go:781]: This is a nice change, but let's get a test for it as well.

**Review by codingllama — COMMENTED** [review]
I get separate file for auth_token, but I'm not sure why I'd want one for CA Pins. CA Pins aren't sensitive, so it looks to me that placing them in a separate file adds surface for undue changes. Curious to hear what other reviewers make of it (@nklaassen @r0mant).

**Comment by r0mant** [comment]
> I get separate file for auth_token, but I'm not sure why I'd want one for CA Pins. CA Pins aren't sensitive, so it looks to me that placing them in a separate file adds surface for undue changes. Curious to hear what other reviewers make of it (@nklaassen @r0mant).

They're not sensitive but it is useful for automation (e.g. you can build an AMI with your cluster's CA pin embedded).

File: lib/service/cfg.go [hunk 3]
```diff
@@ -276,6 +276,26 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
...
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:280]: This is a good point, and on further thinking I realized that if the file itself is empty, that file content would not be "filtered out" as-is. It was the case before that we ignored a single empty ca_pin, but not multiple. I'm not sure why. I'll have to check the prior behavior with multiple empty strings.

File: lib/utils/token.go [hunk 4]
```diff
@@ -18,20 +18,22 @@ package utils
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/utils/token.go:18]: Yes, I missed that. Not sure what we should name it - we could just move the function into lib/utils/utils.go

File: lib/config/configuration.go [hunk 5]
```diff
@@ -346,8 +346,10 @@ func ApplyFileConfig(fc *FileConfig, cfg *service.Config) error {
 
 	// Read in how nodes will validate the CA. A single empty string in the file
 	// conf should indicate no pins.
-	if len(fc.CAPin) > 1 || (len(fc.CAPin) == 1 && fc.CAPin[0] != "") {
-		cfg.CAPins = fc.CAPin
+	if len(fc.CAPin) > 0 {
```
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/config/configuration.go:349]: I just wanted to avoid the make() in ApplyCAPins, we can push the check into the function though. And below I forgot to check the error - adding that too.

File: lib/utils/token.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/utils/token.go:18]: Yep, it's a tricky one to name. `utils.go` has a lot going on already, so maybe `config.go` ?

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/service/cfg.go:280]: My initial reaction is that an empty file is likely a config error, so I'd be tempted to fail in that case, but that also changes past behavior. :/

Let us know what you found out on this one.

[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:280]: So when adding a node with prior behavior, I observed the following behavior:

1. no ca_pin in config: joining node prints a warning about man-in-the-middle attacks without using ca_pin, but it still starts
2. single blank ca_pin: same warning, still starts (as expected since we skip it if there's just one ca_pin entry and it's blank)
3. multiple blank ca_pin entries: error message when starting the node:

```
2022-05-23T16:09:07-07:00 ERRO [PROC:1]    Node failed to establish connection to cluster: invalid format for certificate pin, expected algorithm:pin, object not found. service/connect.go:86
```

Whether we should preserve this behavior, I'm not sure. If we do, then I could at least add a warning or something about blank CA pins. The message could be more helpful at least, if we do keep it like this.

File: lib/config/fileconf.go (hunk 2 above)
Comment (GavinFrazar) [reply lib/config/fileconf.go:781]: I'm going to test this by using a "real looking" windows path, but it won't actually attempt to read the token file since the testing is happening on a linux box. The important thing is we parse the path correctly.

**Review by marcoandredinis — APPROVED** [review]

File: lib/config/fileconf.go (hunk 2 above)
Comment (GavinFrazar) [reply lib/config/fileconf.go:781]:
```go
func TestAuthenticationConfig_Parse_StaticToken(t *testing.T) {
	roles := []string{"Auth", "Node", "Proxy"}
	joinedRoles := strings.Join(roles[:], ",")
	var token StaticToken = StaticToken(fmt.Sprintf("%s:token", joinedRoles))

	provisionToken, err := token.Parse()
	require.NoError(t, err)

	require.Len(t, provisionToken.Roles, len(roles))
	for i := range roles {
		require.Equal(t, string(provisionToken.Roles[i]), roles[i])
	}
	require.Equal(t, provisionToken.Token, "token")

	// test parsing a windows filepath which contains a colon
	windowsTokenPath := "C:\\path\\to\\some\\token\\file"
	token = StaticToken(fmt.Sprintf("%s:%s", joinedRoles, windowsTokenPath))

	provisionToken, err = token.Parse()
	require.NoError(t, err)

	require.Len(t, provisionToken.Roles, len(roles))
	for i := range roles {
		require.Equal(t, string(provisionToken.Roles[i]), roles[i])
	}
	require.Equal(t, provisionToken.Token, windowsTokenPath)
}
```
Added test and verified it passed.

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:280]: So you're saying if we find empty config file values, just log it as a warning and preserve the behavior that the error occurs down the line when it gets the empty CAPin (`ERRO [PROC:1]...`)

File: lib/service/cfg.go [hunk 6]
```diff
@@ -276,6 +276,29 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
...
 }
 
+// ApplyCAPins assigns the given CA pin(s), filtering out empty pins.
+func (cfg *Config) ApplyCAPins(caPins []string) error {
+	if len(caPins) == 0 {
+		return nil
+	}
+	filteredPins := make([]string, 0, len(caPins))
+	for _, pin := range caPins {
+		if pin != "" {
```
Comment (r0mant) [inline lib/service/cfg.go:286]: Nit: I'd invert the condition to `if pin == "" { continue }` and unindent the block below.

**Review by r0mant — APPROVED** [review]
lgtm once remaining @codingllama's comments are addressed.

File: lib/config/fileconf_test.go [hunk 7]
```diff
@@ -332,6 +334,34 @@ func TestAuthenticationSection(t *testing.T) {
...
+	var token StaticToken = StaticToken(fmt.Sprintf("%s:token", joinedRoles))
+
+	provisionToken, err := token.Parse()
+	require.NoError(t, err)
+
+	require.Len(t, provisionToken.Roles, len(roles))
+	for i := range roles {
+		require.Equal(t, string(provisionToken.Roles[i]), roles[i])
+	}
+	require.Equal(t, provisionToken.Token, "token")
```
Comment (codingllama) [inline lib/config/fileconf_test.go:349]: Simplify?

```suggestion
	want := &types.ProvisionTokenV1{
		Roles: []types.SystemRole{
			types.RoleAuth, types.RoleNode, types.RoleProxy,
		},
		Expires: provisionToken.Expires, // or replace with the value you want
		Token: "token",
	}
	require.Equal(t, want, provisionToken)
```

it's easier to read/reason about and has a lot less logic that could go wrong.

File: lib/config/fileconf_test.go [hunk 8]
```diff
@@ -332,6 +334,34 @@ func TestAuthenticationSection(t *testing.T) {
 	}
 }
 
+func TestAuthenticationConfig_Parse_StaticToken(t *testing.T) {
+	roles := []string{"Auth", "Node", "Proxy"}
+	joinedRoles := strings.Join(roles[:], ",")
+	var token StaticToken = StaticToken(fmt.Sprintf("%s:token", joinedRoles))
```
Comment (codingllama) [inline lib/config/fileconf_test.go:340]: No brainer simplifications:

```suggestion
	roles := []string{"Auth", "Node", "Proxy"}
	joinedRoles := strings.Join(roles, ",")
	token := StaticToken(fmt.Sprintf("%s:token", joinedRoles))
```

but I'd much rather do:

```suggestion
	token := StaticToken("Auth,Node,Proxy:token")
```

much easier to read.

File: lib/config/fileconf_test.go [hunk 9]
```diff
@@ -332,6 +334,34 @@ func TestAuthenticationSection(t *testing.T) {
...
+	require.NoError(t, err)
+
+	require.Len(t, provisionToken.Roles, len(roles))
+	for i := range roles {
+		require.Equal(t, string(provisionToken.Roles[i]), roles[i])
+	}
+	require.Equal(t, provisionToken.Token, "token")
+
+	// test parsing a windows filepath which contains a colon
+	windowsTokenPath := "C:\\path\\to\\some\\token\\file"
```
Comment (codingllama) [inline lib/config/fileconf_test.go:352]:
```suggestion
	windowsTokenPath := `C:\path\to\some\token\file`
```

File: lib/config/fileconf_test.go [hunk 10]
```diff
@@ -332,6 +334,34 @@ func TestAuthenticationSection(t *testing.T) {
...
+	token = StaticToken(fmt.Sprintf("%s:%s", joinedRoles, windowsTokenPath))
+
+	provisionToken, err = token.Parse()
+	require.NoError(t, err)
+
+	require.Len(t, provisionToken.Roles, len(roles))
+	for i := range roles {
+		require.Equal(t, string(provisionToken.Roles[i]), roles[i])
+	}
+	require.Equal(t, provisionToken.Token, windowsTokenPath)
```
Comment (codingllama) [inline lib/config/fileconf_test.go:362]: Same comments as above.

Let's make this into a table-drive test, it's a good fit.

https://github.com/golang/go/wiki/TableDrivenTests

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/service/cfg.go:280]: I'm suggesting:

1. Let's make sure ApplyCAPins keeps the behavior of the old `if len > 1`, preferably erroring if any pins are empty in that scenario
2. Let's check the return from TryReadValueAsFile in ApplyCAPins to make sure it's not empty and error if it is.

File: lib/config/fileconf_test.go (hunk 9 above)
Comment (GavinFrazar) [reply lib/config/fileconf_test.go:352]: I appreciate the comments, as you can probably tell I'm still new to Go.

File: lib/config/fileconf_test.go (hunk 10 above)
Comment (GavinFrazar) [reply lib/config/fileconf_test.go]: That simplified things greatly, thanks!

File: lib/utils/config.go [hunk 11]
```diff
@@ -18,20 +18,22 @@ package utils
...
+// otherwise, treat it as a value.
+// It only support absolute paths to avoid ambiguity in interpretation of the value
+func TryReadValueAsFile(value string) (string, error) {
+	if !filepath.IsAbs(value) {
+		return value, nil
 	}
-	// treat it as a file
-	out, err := os.ReadFile(token)
+	// treat it as an absolute filepath
+	out, err := os.ReadFile(value)
```
Comment (nklaassen) [inline lib/utils/config.go:36]: If the user has multiple CA pins and wants them in a file rather than the teleport config, IMO it would be nice to allow multiple pins in the same file rather than requiring a separate file for each pin

Comment (GavinFrazar) [reply lib/utils/config.go:36]: I like that suggestion and the same would apply to tokens

Comment (nklaassen) [reply lib/utils/config.go:36]: I believe you can only have a single `auth_token`, though

Comment (GavinFrazar) [reply lib/utils/config.go:36]: You're right, I was confusing myself with `auth_service.tokens`

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:280]: 1. agreed
2. agreed, but I think we should check for empty file contents in TryReadValueAsFile and log a warning as well. This will provide some hint of what went wrong parsing for any code that goes through this function. It may also hint at a config mistake, even if there is no error thrown,  for example I tested `auth_service.tokens` parsing behavior on master: empty tokens or empty token files show no warning nor error and proceed to start the server. Same for an empty `teleport.auth_token` file:

```
teleport:
  # valid, no warning nor error
  auth_token: /Users/gavin/some-empty-file
...
auth_service:
  tokens:
    # both valid, no warning/error
    - node:/Users/gavin/some-empty-file
    - "node:"
```

3. Take @nklaassen's suggestion to allow multiple values in a file. I can't foresee this breaking any past behavior.

File: lib/config/fileconf_test.go (hunk 8 above)
Comment (GavinFrazar) [reply lib/config/fileconf_test.go]: Simplified the test to be table driven, but for convenience I still do the string concatenation in there. I think this is resolved

File: lib/utils/config.go (hunk 11 above)
Comment (nklaassen) [reply lib/utils/config.go:36]: Oh right, yeah I think that is fine, it should be backward compatible. I think those can include commas, right? Probably best to just split on whitespace then

Comment (GavinFrazar) [reply lib/utils/config.go:36]: The config for those looks like either of these:
```
auth_service:
    tokens:
        - node:some-static-token
        - node,proxy:/Some/Path/To/Token/File
```

So the file is the same format.

If we do allow multiple pins in a file, should they just be newline separated? That would seem the most obvious way to me.

File: lib/service/cfg.go [hunk 12]
```diff
@@ -277,20 +277,25 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
...
-		if pin != "" {
-			var err error
-			pin, err = utils.TryReadValueAsFile(pin)
-			if err != nil {
-				return trace.Wrap(err)
-			}
-			filteredPins = append(filteredPins, pin)
+	for _, pinOrPath := range caPins {
+		if pinOrPath == "" {
+			continue
```
Comment (codingllama) [inline lib/service/cfg.go:288]: Should we error here? We already cleared the "len = 1 && empty" case above, so we know this is a strange looking config (that will likely fail further down the line)?

You could just let it go through TryReadValueAsFile as well, and let it error on line 295.

File: lib/utils/config.go [hunk 13]
```diff
@@ -17,24 +17,32 @@ limitations under the License.
...
+	contents, err := os.ReadFile(value)
 	if err != nil {
 		return "", trace.ConvertSystemError(err)
 	}
 	// trim newlines as tokens in files tend to have newlines
-	return strings.TrimSpace(string(out)), nil
+	out := strings.TrimSpace(string(contents))
+
+	if out == "" {
+		fmt.Fprintf(os.Stderr, "WARNING: empty config value file: %v\n", value)
```
Comment (codingllama) [inline lib/utils/config.go:45]: I would ditch the warning. Either the empty value is allowed and the warning is false, or the empty value is not allowed and the caller should escalate to a proper error.

File: lib/config/fileconf_test.go [hunk 14]
```diff
@@ -332,6 +332,34 @@ func TestAuthenticationSection(t *testing.T) {
...
 
+func TestAuthenticationConfig_Parse_StaticToken(t *testing.T) {
+	t.Parallel()
+
+	tests := []struct {
+		desc  string
+		token string
+	}{
+		{"file path on windows", `C:\path\to\some\file`},
+		{"literal string", "some-literal-token"},
```
Comment (codingllama) [inline lib/config/fileconf_test.go:343]: nit: use field names, it's more readable and more resistant to future changes.

```suggestion
		{desc: "file path on windows", token: `C:\path\to\some\file`},
		{desc: "literal string", token: "some-literal-token"},
```

(Sadly, the reference I pasted for table tests doesn't do this.)

File: lib/config/fileconf_test.go [hunk 15]
```diff
@@ -332,6 +332,34 @@ func TestAuthenticationSection(t *testing.T) {
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			authToken := StaticToken("Auth,Node,Proxy:" + tt.token)
+			provisionToken, err := authToken.Parse()
+			require.NoError(t, err)
```
Comment (codingllama) [inline lib/config/fileconf_test.go:349]: This looks much better, thanks.

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/service/cfg.go:281]: Partly agree with 2., see other comments.

I don't see 3. implemented?

File: lib/service/cfg.go [hunk 16]
```diff
@@ -276,6 +276,34 @@ func (cfg *Config) ApplyToken(token string) (bool, error) {
...
+		}
+		// an empty pin file is less obvious than a blank ca_pin in the config yaml.
+		if pin == "" {
+			return trace.BadParameter("empty ca_pin file: %v", pinOrPath)
+		}
+		filteredPins = append(filteredPins, pin)
+	}
+	if len(filteredPins) == 0 {
+		return nil
+	}
```
Comment (codingllama) [inline lib/service/cfg.go:302]:
```suggestion
```

The if doesn't do much, the end result is the same if it goes through. Plus, if you take the comment above, len=0 shouldn't happen here.

File: lib/service/cfg.go (hunk 12 above)
Comment (GavinFrazar) [reply lib/service/cfg.go:288]: The idea was that empty ca_pin values are fine, but empty ca_pin files are probably not expected and should produce an error. It's a little strange that prior behavior would allow one and only one empty ca_pin but not allow it if there are more than one. I suppose if we kept this then we could remove the "if len == 1" check above though.

This won't break any existing teleport configs if we allow empty values in the yaml. I don't have a strong preference though, so we can error here if you prefer that.

File: lib/utils/config.go (hunk 13 above)
Comment (GavinFrazar) [reply lib/utils/config.go:45]: If teleport.auth_token or auth_service.tokens have a path to an empty file it's currently allowed, but may be surprising to a user having problems getting their node to join a cluster. The idea is to make non-obvious config mistakes obvious to a user, so I'd like to print some kind of message telling them/warning them about it.

Assuming we do want to point out a non-obvious mistake like this, I can't see a case where the caller wouldn't want them to be at least notified - only cases where we decide to escalate to an error as well.

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:281]: I didn't implement 3. yet, as I wasn't sure if everyone was on board with that

File: lib/service/cfg.go (hunk 12 above)
Comment (GavinFrazar) [reply lib/service/cfg.go:288]: > Let's make sure ApplyCAPins keeps the behavior of the old if len > 1, preferably erroring if any pins are empty in that scenario

My bad, I totally agreed with you above and then submitted code which doesn't error for empty pins. During a 1x1 with Roman, we discussed this and decided on allowing empty ca_pin values if it's directly in the yaml and error only for an empty ca_pin file. We agreed on that because it's more consistent blank ca_pin yaml behavior and wouldn't break any existing configs.

File: lib/utils/config.go (hunk 13 above)
Comment (codingllama) [reply lib/utils/config.go:45]: Got it.

Are you avoiding `log.Warnf` here because we think it might not be properly initialized? If yes please add a comment, otherwise let's move to `log.Warnf`.

Comment (GavinFrazar) [reply lib/utils/config.go:45]: I actually wasn't sure how to handle it. I searched our code and found we use both fmt and logging to print warnings. More locally though, we do use log in the utils package, so I can do that

File: lib/utils/config.go (hunk 11 above)
Comment (nklaassen) [reply lib/utils/config.go]: Yeah `bufio.Scanner` splits by line by default, might as well do that

File: lib/service/cfg.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (GavinFrazar) [reply lib/service/cfg.go:281]: 3. Implemented and I updated tests for config value filepath in tokens/pins

File: lib/utils/config.go (hunk 11 above)
Comment (GavinFrazar) [reply lib/utils/config.go]: I kept it more generic and just have the TryReadValueAsFile function return a string. Then on caller side I use strings.Split

**Review by codingllama — APPROVED** [review]
